src/utils/graph_utils.py

```diff
@@ -0,0 +1,137 @@
+            file['sum_elements'][element] += 1
+
+        #Average by the number of contributing elements
+        for node_feature,_ in features:
+             file[node_feature] = file[node_feature]/file['sum_elements']
+       
+
+    return bin_files
+
+
+
+def bin_to_torchGraph(
+        bin_files: list[dict],
+        features_list: list[str],
+        target_idx: int
+):
+
+    """
+    Funtion to create torch graph "Data" structures from nextsim bin output files (as .npz)   
+    For each file 3 torch.tensors are computed and used to create the data object:
+        node features [num_nodes, num_node_features]
+        edges [2, num_edges]
+        edges atributtes [num_edges, num_edge_features]
+
+    Arguments:
+        bin_files: list[dict]
+            list of dictionaries containing nextsim outputs
+        features: list[str]
+            list of features to include on each node
+        target_idx: int
+            index of the target node
+        
+    returns:
+        grads : list[torch_geometric.data.Data]
+            list containing torch geometric data objetcs
+            
+    """
+    graph_list=[]
+
+    for idx,hour_graph in tqdm(enumerate(bin_files),"Converting bins to torch graphs..."):
+
+        #get the next pos of target node
+        target_coords = torch.tensor([bin_files[idx+1]['x'][target_idx],bin_files[idx+1]['y'][target_idx]])
+
+        #concat all node features sequentially (following index(t) number) in a tensor
+        features = []
+        idx_list = [] #keep track of index inside node feature tensor
+        for key,item in hour_graph.items():
+
+            if key in features_list:
+                idx_list.append(key)
+                features.append(torch.tensor(np.array([item])))
+
+        if len(features)>0:
+            node_features = torch.cat(features).t().to(torch.float32)
+        else:
+            raise ValueError("None of the features specified is on the graph")
+
+        #find all distinct (undirected) edges from every triangle
+        edges = []
+        for triangle in hour_graph['t']:
+            edges += [ tuple(triangle[:2]), tuple(triangle[1:]), tuple(triangle[[0,-1]]) ] #tuples since we set() it later
+
+        # Get a unique set and convert it to a torch tensor
+        edges = torch.tensor(list(set(edges))).t()
+
+        #Now we need to consult x,y coordinates of each node of the edges and compute the edge distance
+        # for each each row of edge ends we retrieve this info by index
+        # and we stack it as a 2xE (2 for each edge end, E as number of edges)
+        coord_idx= [i for i,key in enumerate(idx_list) if key in ['x','y']]
+        if len(coord_idx)==2:
+            edges_coordinates = [
+                torch.stack(
+                    [
+                        node_features[edge_row][:,coord_idx[0]],
+                        node_features[edge_row][:,coord_idx[1]]
+                    ]
+                )
```

Comment:
Maybe vectorisation can be used here as well? I don't know if toch tensor allows that. Numpy does:
`node_features[edges, coord_idx[0]]`

Reply:
I can not see the way here right now. Torch works as numpy in 99% of the cases :) 
In here we want to index with tuples to retrieve x,y in both ends, if we would use the whole edge tensor (2,n_edges) the indexing wouldn't make sense, I think. There might be another way, but cant see it now, my mind still works in "slow motion" with graphs on tensors. hehe